cloud/ec2_fleet.go

```diff
@@ -0,0 +1,538 @@
+			{
+				AssociatePublicIpAddress: aws.Bool(true),
+				DeviceIndex:              aws.Int64(0),
+				Groups:                   ec2Settings.getSecurityGroups(),
+				SubnetId:                 aws.String(ec2Settings.SubnetId),
+			},
+		}
+		if ec2Settings.IPv6 {
+			launchTemplate.NetworkInterfaces[0].SetIpv6AddressCount(1).SetAssociatePublicIpAddress(false)
+		}
+	} else {
+		launchTemplate.SecurityGroups = ec2Settings.getSecurityGroups()
+	}
+
+	userData, err := bootstrapUserData(ctx, evergreen.GetEnvironment(), h, ec2Settings.UserData)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "could not add bootstrap script to user data")
+	}
+	ec2Settings.UserData = userData
+
+	if ec2Settings.UserData != "" {
+		var expanded string
+		expanded, err = expandUserData(ec2Settings.UserData, m.settings.Expansions)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "problem expanding user data")
+		}
+		userData := base64.StdEncoding.EncodeToString([]byte(expanded))
+		launchTemplate.UserData = &userData
+	}
+
+	createTemplateResponse, err := m.client.CreateLaunchTemplate(ctx, &ec2.CreateLaunchTemplateInput{
+		LaunchTemplateData: launchTemplate,
+		// mandatory field may only contain letters, numbers, and the following characters: - ( ) . / _
+		LaunchTemplateName: aws.String(fmt.Sprintf("%s", util.RandomString())),
+	})
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "can't upload config template to AWS")
+	}
+	err = validateEc2CreateTemplateResponse(createTemplateResponse)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "invalid create template response")
+	}
+
+	return createTemplateResponse.LaunchTemplate.LaunchTemplateId, createTemplateResponse.LaunchTemplate.LatestVersionNumber, nil
+}
+
+func (m *ec2FleetManager) requestFleet(ctx context.Context, ec2Settings *EC2ProviderSettings, templateID *string, templateVersion *int64) (*string, error) {
+	var overrides []*ec2.FleetLaunchTemplateOverridesRequest
+	var err error
+	if ec2Settings.VpcName != "" {
+		overrides, err = m.makeOverrides(ctx, ec2Settings)
+		if err != nil {
+			return nil, errors.Wrapf(err, "can't make overrides for VPC '%s'", ec2Settings.VpcName)
+		}
+	}
+
+	// Create a fleet with a single spot instance from the launch template
+	createFleetInput := &ec2.CreateFleetInput{
+		LaunchTemplateConfigs: []*ec2.FleetLaunchTemplateConfigRequest{
+			{
+				LaunchTemplateSpecification: &ec2.FleetLaunchTemplateSpecificationRequest{
+					LaunchTemplateId: templateID,
+					Version:          aws.String(strconv.Itoa(int(*templateVersion))),
+				},
+				Overrides: overrides,
+			},
+		},
+		TargetCapacitySpecification: &ec2.TargetCapacitySpecificationRequest{
+			TotalTargetCapacity:       aws.Int64(1),
+			DefaultTargetCapacityType: aws.String(ec2.DefaultTargetCapacityTypeSpot),
+		},
+		Type: aws.String(ec2.FleetTypeInstant),
+	}
+	createFleetResponse, err := m.client.CreateFleet(ctx, createFleetInput)
```

Comment:
Does this create 1 fleet request per host? I had an issue in my original attempt at this where there's a hard limit of 1000 active fleets in a region at a time (https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-fleet.html#EC2-fleet-limits), which would effectively limit evergreen to 1000  running hosts.

Reply:
I didn't realized you had worked on this. (I would have asked you a lot of questions)

It does create a single host per Fleet. We were wondering about this and AWS assured us that there is no limit and that they needed to change their docs.